fvm/src/gas/outputs.rs

```diff
@@ -61,27 +59,106 @@ impl GasOutputs {
     }
 }
 
-fn compute_gas_overestimation_burn(gas_used: i64, gas_limit: i64) -> (i64, i64) {
-    const GAS_OVERUSE_NUM: i64 = 11;
-    const GAS_OVERUSE_DENOM: i64 = 10;
+fn compute_gas_overestimation_burn(gas_used: u64, gas_limit: u64) -> (u64, u64) {
+    const GAS_OVERUSE_NUM: u128 = 11;
+    const GAS_OVERUSE_DENOM: u128 = 10;
 
     if gas_used == 0 {
         return (0, gas_limit);
     }
 
-    let mut over = gas_limit - (GAS_OVERUSE_NUM * gas_used) / GAS_OVERUSE_DENOM;
-    if over < 0 {
-        return (gas_limit - gas_used, 0);
-    }
+    // Convert to u128 to prevent overflow on multiply.
+    let gas_used = gas_used as u128;
+    let gas_limit = gas_limit as u128;
+
+    // This burns (N-10)% (clamped at 0% and 100%) of the remaining gas where N is the
+    // overestimation percentage.
+    let over = gas_limit
+        .saturating_sub((GAS_OVERUSE_NUM * gas_used) / GAS_OVERUSE_DENOM)
+        .min(gas_used);
+
+    // We handle the case where the gas used exceeds the gas limit, just in case.
+    let gas_remaining = gas_limit.saturating_sub(gas_used);
+
+    // This computes the fraction of the "remaining" gas to burn and will never be greater than 100%
+    // of the remaining gas.
+    let gas_to_burn = (gas_remaining * over) / gas_used;
+
+    // But... we use saturating sub, just in case.
+    let refund = gas_remaining.saturating_sub(gas_to_burn);
 
-    if over > gas_used {
-        over = gas_used;
+    (refund as u64, gas_to_burn as u64)
+}
+
+// Adapted from lotus.
+#[test]
+fn overestimation_burn_test() {
```

Comment:
Out of curiosity -- would it be straightforward to do a one-off run of these tests against the old logic just to make sure we haven't inadvertently introduced a regression?

Reply:
I just stole these tests from lotus, which have been running against the lotus implementation of this function.